### PR TITLE
authz: Allow directory read access for sub repo permissions

### PR DESCRIPTION
**Pull request by ryanslade**
If a path is allowed by sub repo permissions, we should also allow read
access to any directories the lie above that path in the tree.

For example, if we allow access to "foo/bar/data.txt" we should also
allow the path "foo/" and "foo/bar/" so that we can navigate down the
tree.

This does NOT imply access to all items living under "foo/" or
"foo/bar/"

**Caveat**: The logic above cannot be applied to rules that start with a wildcard
because it can lead to allowing read access to *any* directory.

## Test plan

More test cases added
Tested manually

### CODE REVIEW

**Comment by ryanslade** [comment]
@mollylogue PTAL

I'm hoping this will fix the file browsing issue we have with sub-repo permissions.

File: internal/authz/sub_repo_perms_test.go [hunk 1]
```diff
@@ -222,6 +222,96 @@ func TestCanReadAllPaths(t *testing.T) {
...
+		},
+		{
+			pathIncludes:  []string{"foo/bar/"},
+			canReadAll:    []string{"foo/", "foo/bar/"},
+			cannotReadAny: []string{"foo/thing.txt", "foo/bar/thing.txt"},
+		},
+		{
+			pathIncludes:  []string{"**/foo/bar/thing.txt"},
+			canReadAll:    []string{"a/b/foo/bar/", "a/foo/", "a/foo/bar/"},
+			cannotReadAny: []string{"foo/", "foo/bar/"},
```
Comment (ryanslade) [inline internal/authz/sub_repo_perms_test.go:260]: Added this failing test that I'll look into on Monday

**Review by sashaostrikov — APPROVED** [review]
LGTM!

**Review by mollylogue — APPROVED** [review]
🙏 Looks good. Thanks for tackling this. Should we document somewhere the case with the rule beginning w/ a wildcard? The comment in code is probably sufficient, just wondering if it's worth surfacing elsewhere (or if it would just create more noise)

**Comment by ryanslade** [comment]
> 🙏 Looks good. Thanks for tackling this. Should we document somewhere the case with the rule beginning w/ a wildcard? The comment in code is probably sufficient, just wondering if it's worth surfacing elsewhere (or if it would just create more noise)

I think it may be too noisy. We'd have to explain the the edge case in our docs and also explain that it would only happen if no other rules applies. I think in practice it's going to be be rare. I'll take a look at the sample protection tables we have.

**Pull request merged**